app/models/manageiq/providers/redhat/infra_manager.rb

```diff
@@ -115,16 +122,51 @@ def prepare_disk(disk_spec, ems_storage_uid)
     }
   end
 
-  # RHEVM requires that the memory of the VM will be bigger or equal to the reserved memory at any given time.
-  # Therefore, increasing the memory of the vm should precede to updating the reserved memory, and the opposite:
-  # Decreasing the memory to a lower value than the reserved memory requires first to update the reserved memory
   def update_vm_memory(rhevm_vm, memory)
-    if memory > rhevm_vm.attributes.fetch_path(:memory)
-      rhevm_vm.memory = memory
-      rhevm_vm.memory_reserve = memory
+    # Get the current state of the virtual machine, and the current amounts of memory:
+    attributes = rhevm_vm.attributes
+    current_state = attributes.fetch_path(:status, :state)
+    current_memory = attributes.fetch_path(:memory)
+    current_guaranteed = attributes.fetch_path(:memory_policy, :guaranteed)
+
+    # If the virtual machine is running then the difference in memory has to be a multiple of 256 MiB, otherwise
+    # the engine will not perform the hot plug of the new memory. The reason for this is that hot plugging of
+    # memory is performed adding a new virtual DIMM to the virtual machine, and the size of the virtual DIMM
+    # is 256 MiB. This means that we need to round the difference up to the closest multiple of 256 MiB.
+    if current_state == 'up'
+      delta = memory - current_memory
+      remainder = delta % HOT_PLUG_DIMM_SIZE
+      if remainder > 0
+        adjustment = HOT_PLUG_DIMM_SIZE - remainder
+        delta += adjustment
+        memory = current_memory + delta
+        _log.info(
+          "The change in virtual memory needs to be a multiple of #{HOT_PLUG_DIMM_SIZE / 1.megabyte} MiB, " +
+          "so it will be adjusted to #{memory / 1.megabyte} MiB."
+        )
+      end
+    end
+
+    # The engine requires that the virtual memory is bigger or equal than the guaranteed memory at any given
+    # time. Therefore, we need to adjust the guaranteed memory so that it is the minimum of the previous
+    # guaranteed memory and the new virtual memory.
+    if current_guaranteed <= memory
+      guaranteed = current_guaranteed
+    else
+      guaranteed = memory
+      _log.info(
+        "The guaranteed physical memory needs to be less or equal than the virtual memory, so it will be " +
+        "adjusted to #{guaranteed / 1.megabyte} MiB."
+      )
+    end
```

Comment:
lines 150-161 can be extracted to a method which is responsible for calculating the guaranteed memory